proposals/1533-host-specific-tasklist.md

```diff
@@ -0,0 +1,241 @@
+
+## Implementation
+
+### Session Model
+
+All sessions are **resource specific**. It basically means a session is always tied to some kind of resource (not the worker). The resource can be CPU, GPU, memory, file descriptors etc and we want all the activities within a session to be executed by the same worker which owns the resource. If worker dies and restarts, as long as it owns the same resource, we can try to reestablish the session (this is future work). Also user needs to ensure that resource is owned by only one worker.
+
+Note that when creating a session, user doesn’t need to specify the resource that is consumed by the session. This is because right now one worker owns only one resource, so there’s only one resource type on the worker. As a result, as long as the user specifies the correct tasklist in the context passed into the createSession API, the correct type of resource will be consumed (since there’s only one). Later, a worker can support multiple types of resources and at that time, users need to specify the resource type when creating a session.
+
+### Workflow Context
+
+When user calls the CreateSession() or CreateSessionWithHostID() API, a structure contains information about the created session will be stored in the returned context. The structure contains three two pieces of information:
+
+1. **SessionID**: a unique ID for the created session. (Exported)
+
+2. **tasklist**: the resource specific tasklist used by this session. (Not Exported)
+
+3. **state:** state of the session (Not Exported).  It can take one of the three values below:
+
+   1. **Open**: when the session is created and running
+   2. **Failed**: when worker is down and session can’t continue
+   3. **Closed**: when user closes the session through the API call and session is successfully closed.
+
+When user calls the CompleteSession() API, the state of the session will be changed to "Closed" in place, that’s why this API won’t return a new context variable.
+
+User has no way to set the state of a session to failed. The state of a session is "failed" only when the worker executing the session is down and the change from “open” to “failed” is done in the background. Also notice that it’s possible that the worker is down, but the session state is still “Open” (since it takes some time to detect the worker failure). In this case, when a user executes an activity, the activity will still be scheduled on the resource specific tasklist. However, since no one will poll from this task list, that activity will timeout and user will get an error.
+
+Note that If user tries to call CreateSession() with a context which already has an open session, the call will return an error. Same thing happens if a call to CompleteSession() is made when there’s no open session. You can call it with a failed session though, but it won’t do anything. This allowed because the state transition to "failed" is done in the background, and user may not know the session has failed.
```

Comment:
Note that If user -> Note that if a user
context which already -> context that already
Same thing -> The same thing
remove "though"
This allowed -> This is allowed